Introduce path and deprecate hub_con and config_tasks as args to submission_tmpl

annakrystalli opened the pull request:
This PR resolves #165 and in the process also resolves #137 

The PR introduces `hub_path` as main argument to `submission_tmpl()` and deprecates arguments `hub_con` and `config_tasks` (#165 & #137). This way, all that is required by the user to create a submission template is the path to a hub directory.

The change is currently back-compatible as the old arguments still work while issuing a deprecation warning. The arguments will be fully removed in a later version (see #197).

Tests were also updated to work with hub_paths while some tests to ensure current back-compatibility were retained. Note that this change requires a fully configured hub to work, but tests that are based on standalone config files have been accomodated by using local mocking of the `read_config()` function now used internally to read config files.

zkamvar commented on R/submission_tmpl.R:
```diff
@@ -265,3 +251,34 @@ subset_complete_cases <- function(tmpl_df) {
+      checkmate::assert_list(config_tasks)
+    },
+    hub_path = {
+      checkmate::assert_directory_exists(hub_path)
+      read_config(hub_path)
```
I often hear from modelers that they do not often work directly with the hub on their machine when creating their model outputs. 

If we allow for passing individual config files to be passed to `hub_path` (which could be renamed to `path` to avoid confusion), then modelers could work where they are comfortable working and they could even pass the URL for the config. 

```suggestion
      if (utils::file_test("-d", hub_path)) {
        checkmate::assert_directory_exists(hub_path)
        read_config(hub_path)
      } else {
        checkmate::assert_file_exists(hub_path, extension = "json")
        read_config_file(hub_path)
      }
```

zkamvar reviewed (changes requested):
Thank you! I think the ergonomics of this function are much better with a single argument. It's a much simpler solution than my suggestion from #137. Your method of using a function to switch out the deprecated arguments like that is a nice touch!

I do have a request for an improvement: allow the first argument to take either a hub or a config file and change the argument name to be `path`. 

My rationale is that it improves testing and modeler ergonomics:

> tests that are based on standalone config files have been accommodated by using local mocking of the `read_config()` function now used internally to read config files.

If we allow for a config file as the first argument as well, then these local mocks would not be necessary.

I've described the modeler ergonomics in my comment below.

annakrystalli commented:
Good idea! Not being able to create a template from a config file was bothering me too. I was just trying to keep the interface as simple as possible. This solution is a good way round it.

I've implemented with minor adaptations. Specifically I ensure we check that `path` exists first before trying to determine whether it's a directory or file. I also use an `fs` function (which is already an import) as it's name is more self descriptive.

I've also updated the test to remove local bindings, done a bit of reorganising for clarity and added a few more specific tests that were being tested in a more high-level fashion but not clearly indicated as such.


Re:
> I often hear from modelers that they do not often work directly with the hub on their machine when creating their model outputs.

This is interesting and makes me wonder, how to they ensure they have access to an up to date version of the config locally?

zkamvar reviewed (changes requested):
Thank you for the updates! I have two final suggestions:

1. update the NEWS item
2. add a catch so that non-json files throw an error

to answer your question:

> > I often hear from modelers that they do not often work directly with the hub on their machine when creating their model outputs.
> This is interesting and makes me wonder, how to they ensure they have access to an up to date version of the config locally?

That's a good question, but they wouldn't necessarily need to do this if they have an internet connection because they could do something like this:

```r
cfg <- "https://raw.githubusercontent.com/reichlab/variant-nowcast-hub/refs/heads/main/hub-config/tasks.json"
```

annakrystalli commented on tests/testthat/_snaps/submission_tmpl.md:
```diff
@@ -163,17 +177,26 @@
        4 2022-12-26    wk ahead inc~       1 01       mean        <NA>              NA
        5 2022-12-26    wk ahead inc~       2 02       mean        <NA>              NA
        6 2022-12-26    wk ahead inc~       1 02       mean        <NA>              NA
-       7 2022-12-26    wk ahead inc~       2 US       sample      1                 NA
-       8 2022-12-26    wk ahead inc~       2 01       sample      1                 NA
-       9 2022-12-26    wk ahead inc~       2 02       sample      1                 NA
-      10 2022-12-26    wk ahead inc~       1 US       sample      2                 NA
+       7 2022-12-26    wk ahead inc~       2 US       sample      s1                NA
+       8 2022-12-26    wk ahead inc~       1 US       sample      s2                NA
+       9 2022-12-26    wk ahead inc~       2 01       sample      s3                NA
+      10 2022-12-26    wk ahead inc~       1 01       sample      s4                NA
```
It's because the checks have been re-organised, some removed hence some new snapshots have been made

annakrystalli commented on R/submission_tmpl.R:
```diff
@@ -265,3 +258,39 @@ subset_complete_cases <- function(tmpl_df) {
+      if (fs::is_dir(path)) {
+        return(read_config(path))
+      }
+      read_config_file(path)
```
I'm tempted to say this need to be part of r`ead_config_file()`, hence deserves a new issue in `hubUtils` instead

annakrystalli replied:
See https://github.com/hubverse-org/hubUtils/issues/209

annakrystalli commented:
> That's a good question, if they have an internet connection, they could do something like this and get the file since jsonlite::fromJSON() understands URLs
>
> cfg <- "https://raw.githubusercontent.com/reichlab/variant-nowcast-hub/refs/heads/main/hub-config/tasks.json"
submission_tmpl(path = cfg, ...)


This won't work unfortunately with the current setup:
``` r
library(hubValidations)
path <- "https://raw.githubusercontent.com/reichlab/variant-nowcast-hub/refs/heads/main/hub-config/tasks.json"
submission_tmpl(
  path,
  round_id = "2025-01-25"
  )
#> Error in `submission_tmpl()`:
#> ! `path`
#>   'https://raw.githubusercontent.com/reichlab/variant-nowcast-hub/refs/heads/main/hub-config/tasks.json'
#>   does not exist.
```

<sup>Created on 2025-02-04 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

zkamvar reviewed (approved):
> This won't work unfortunately with the current setup:

Oh yeah, that's a good point. The `fs::file_exists()` catches that and it doesn't know about URLs :/

annakrystalli commented:
Just a heads up @zkamvar 

Given the changes proposed in https://github.com/hubverse-org/hubUtils/pull/210 which will solve a couple of the outstanding issues here, I'm going to wait till we merge that, then adapt and re-request review. 🚀

annakrystalli commented:
Ready for review @zkamvar !

zkamvar commented on R/submission_tmpl.R:
```diff
@@ -265,3 +300,57 @@ subset_complete_cases <- function(tmpl_df) {
+    },
+    path = {
+      # Read config from cloud hub
+      if (inherits(path, "SubTreeFileSystem")) {
+        if (hubUtils::is_s3_base_fs(path)) {
+          return(read_config(path))
+        }
+        return(read_config_file(path))
+      }
+      # Read config file from URL
+      if (hubUtils::is_url(path)) {
+        if (hubUtils::is_github_repo_url(path)) {
+          return(read_config(path))
+        }
+        return(read_config_file(path))
+      }
+      # Read local config
+      if (!fs::file_exists(path)) {
+        cli::cli_abort("{.arg path} {.file {path}} does not exist.",
+          call = rlang::caller_env(1)
+        )
+      }
+      if (fs::is_dir(path)) {
+        return(read_config(path))
+      }
+      read_config_file(path)
```
Since much of the heavy lifting is done inside `read_config()` and `read_config_file()`, the main task is to determine if we have a directory or a file. This can be achieved with two indicator variables. 


```suggestion
      is_s3_dir <- inherits(path, "SubTreeFileSystem") && hubUtils::is_s3_base_fs(path)
      is_local_dir <- is.character(path) && fs::path_ext(path) == ""
      if (is_s3_dir || is_local_dir) {
        read_config(path)
      } else {
        read_config_file(path)
      }
```

annakrystalli replied:
‼ This will not pass tests until https://github.com/hubverse-org/hubUtils/pull/211 is merged

Nice simplification! There however exceptions that this wasn't handling well. For example, if someone erroneously supplied `"https://github.com/hubverse-org/schemas/tree/main/v5.0.0"` as a path, it was identified as path and `read_config_file()` was invoked. That's why I added an additional check which catches that and returns a more informative error message.

`is.character(path) && fs::path_ext(path) == ""` is still a bit hacky as if a URL like `"https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/README"` is supplied, it would trigger `read_config()` erroneously. Unfortunately the only method I found for ensuring the correct read function is deployed requires checking the `Content-Type` to see if it’s a file. I've refrained from adding such a function but have added a test for that exception that would be nice to handle better at some point (now it just fails because it interprets the URL as a folder and creates a path to tasks.json that does not exists).

zkamvar replied:
> Nice simplification! There are however exceptions that this wasn't handling well. For example, if someone erroneously supplied `"https://github.com/hubverse-org/schemas/tree/main/v5.0.0"` as a path, it was identified as a file and `read_config_file()` was invoked. That's why I added an additional check which catches that and returns a more informative error message.
>
> `is.character(path) && fs::path_ext(path) == ""` is still a bit hacky as if a URL like `"https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/README"` or `"https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/LICENSE"` is supplied, it would trigger `read_config()` erroneously.

These are fair points and I definitely agree that the directory check is hacky, but I take solace in the hope that these situations are vanishingly rare and are easily fixed.

zkamvar reviewed (changes requested):
Thank you! I have some requested changes for the news and to simplify the control structure for the `path` argument.

annakrystalli commented on NEWS.md:
```diff
@@ -1,5 +1,8 @@
 # hubValidations (development version)
 
+* Introduced `path` as main argument to `submission_tmpl()` and deprecated arguments `hub_con` and `config_tasks` (#165 & #137). This way, all that is required by the user to create a submission template is the path to a hub directory or `tasks.json` config file. We also added functionality to enable sourcing config files from a
+cloud hub or directly from GitHub which means users do not required a local copy of the hub to create a submission template.
```
Done in 2a8c3a8d057c05f6a5dd0956e9641196b2e3aa7b

zkamvar reviewed (approved):
LGTM!
